src/_pytest/recwarn.py

```diff
@@ -303,6 +328,8 @@ def found_str():
                     for r in self:
                         if issubclass(r.category, self.expected_warning):
                             if re.compile(self.match_expr).search(str(r.message)):
```

Comment:
We might want to extract this to a new method `def matches(self, warning) -> bool:` to share the logic between here and the re-emit function.